osquery/tables/system/darwin/fsevents.cpp

```diff
@@ -0,0 +1,255 @@
+        &fsevent_data[file_path.size() + 1 + sizeof(event_id) + size + offset],
+        sizeof(flags));
+    for (const auto& flag : kFlags) {
+      if (flags & flag.first) {
+        all_flags += flag.second + ",";
+      }
+    }
+
+    all_flags.pop_back();
+    if (sig == kDiskLoggerV1) {
+      size += path_length + 1 + sizeof(event_id) + sizeof(flags);
+      if (boost::starts_with(file_path, "/")) {
+        r["path"] = file_path;
+      } else {
+        r["path"] = "/" + file_path;
+      }
+
+      r["event_id"] = BIGINT(event_id);
+      r["source"] = filename;
+      r["flags"] = all_flags;
+      results.push_back(r);
+      continue;
+    }
+
+    long long node_id = 0;
+    memcpy(&node_id,
+           &fsevent_data[file_path.size() + 1 + sizeof(event_id) +
+                         sizeof(flags) + size + offset],
+           sizeof(node_id));
+    size +=
+        path_length + 1 + sizeof(node_id) + sizeof(event_id) + sizeof(flags);
+    if (boost::starts_with(file_path, "/")) {
+      r["path"] = file_path;
+    } else {
+      r["path"] = "/" + file_path;
+    }
+    r["event_id"] = BIGINT(event_id);
+    r["node_id"] = BIGINT(node_id);
+    r["source"] = filename;
+    r["flags"] = all_flags;
+    results.push_back(r);
+  }
+
+  return size;
+}
+
+void parseFsEventData(const std::vector<char>& fsevent_data,
+                      const std::string& filename,
+                      QueryData& results) {
+  FsEventData event_data;
+
+  unsigned long fsevent_size = 0;
+  while (fsevent_size < fsevent_data.size()) {
+    // Get header data, contains size of stream
+    memcpy(&event_data, &fsevent_data[fsevent_size], kHeaderSize);
+    if ((event_data.sig == kDiskLoggerV2) ||
+        (event_data.sig == kDiskLoggerV1)) {
+      int size = parseFsEvent(fsevent_data,
+                              event_data.stream_size,
+                              event_data.sig,
+                              filename,
+                              fsevent_size,
+                              results);
+      fsevent_size += size;
+
+    } else {
+      LOG(WARNING) << "Unknown FS Event signature: " << event_data.sig;
+      break;
+    }
+  }
+}
+
+void parseEvents(const std::string& file, QueryData& results) {
+  std::ifstream compressed_file(file,
+                                std::ios_base::in | std::ios_base::binary);
+  try {
+    boost::iostreams::filtering_stream<boost::iostreams::input> decompress;
+    decompress.push(boost::iostreams::gzip_decompressor());
+    decompress.push(compressed_file);
+
+    std::vector<char> decompress_data(
+        (std::istreambuf_iterator<char>(decompress)),
+        (std::istreambuf_iterator<char>()));
+    compressed_file.close();
+
+    parseFsEventData(decompress_data, file, results);
+  } catch (const std::exception& err) {
+    LOG(WARNING) << "Failed to parse fsevent file, need to be root: "
+                 << err.what();
+  }
+}
+
+QueryData genFsevents(QueryContext& context) {
```

Comment:
A mandatory `WHERE` clause should be implemented to limit the scope of the query and avoid watchdog issues with simple `SELECT * FROM table` statements